docs/internals/layout_in_storage.rst

```diff
@@ -53,30 +64,56 @@ Mappings and Dynamic Arrays
 
 .. _storage-hashed-encoding:
 
-Due to their unpredictable size, mapping and dynamically-sized array types use a Keccak-256 hash
-computation to find the starting position of the value or the array data.
-These starting positions are always full stack slots.
+Due to their unpredictable size, mappings and dynamically-sized array types cannot be stored
+"in between" the state variables preceding and following them.
+Instead, they are considered to occupy only 32 bytes with regards to the
+:ref:`rules above <storage-inplace-encoding>` and the actual data of mappings and
+dynamcically-sized arrays is stored starting at a different
+storage slot that is computed using a Keccak-256 hash.
 
-The mapping or the dynamic array itself occupies a slot in storage at some position ``p``
-according to the above rule (or by recursively applying this rule for
-mappings of mappings or arrays of arrays). For dynamic arrays,
+Assume the storage location of the mapping or array ends up being a slot ``p``
+after applying :ref:`the storage layout rules <storage-inplace-encoding>`.
+For dynamic arrays,
 this slot stores the number of elements in the array (byte arrays and
 strings are an exception, see :ref:`below <bytes-and-string>`).
-For mappings, the slot is unused (but it is needed so that two equal mappings after each other will use a different
-hash distribution). Array data is located at ``keccak256(p)`` and the value corresponding to a mapping key
-``k`` is located at ``keccak256(k . p)`` where ``.`` is concatenation. If the value is again a
-non-elementary type, the positions are found by adding an offset of ``keccak256(k . p)``.
-
-So for the following contract snippet
-the position of ``data[4][9].b`` is at ``keccak256(uint256(9) . keccak256(uint256(4) . uint256(1))) + 1``::
-
+For mappings, the slot stays empty, but it is needed so that two mappings after each other will use a
+different hash distribution to store their data due to a different starting slot.
+
+Array data is located starting at ``keccak256(p)`` and it is laid out in the same way as
+statically-sized array data would: One element after the other, potentially sharing
+storage slots if the elements are shorter than 32 bytes. Dynamic arrays of dynamic arrays apply this
+rule recursively. The location of ``x[i][j]`` for ``x`` being ``uint32[][]`` is
+computed as follows (again, assuming ``x`` itself is stored at slot ``p``):
+The slot is ``keccak256(keccak2569(p) + i) + floor(j / 8)`` (8 items per slot) and
+the element can be obtained from the slot data ``v`` using ``(v >> ((j % 8) * 2)) & 0xffff``.
```

Comment:
Let's say the uint32 values are 1, 2, 3. Each value is stored in 4 byte groups (32-bit) so the last 12 bytes look like `0x000100020003.` So, isn't the formula to obtain data `(v >> ((j % 8) * 8)) & 0xffffffff`?

Reply:
Right! Is the forumla useful?

Reply:
Depends. If an "advanced" user wants to exploit this to reduce gas usage, maybe?

Reply:
It's useful, I'm trying to look at internal contract state /nonpublic storage on chain so understanding that formula is necessary if you want to introspect values smaller than 32bytes

Reply:
Hi I have a question.

what does `v` in ``(v >> ((j % 8) * 2)) & 0xffff`` refer to?

Also, why are we &ing with 0xffff? Is not anding with 1111 same as the value that we are anding?